### PR TITLE
BE: RBAC: Impl Active Directory populator

### LINKED ISSUE #54: RBAC: Support Active Directory
LdapSecurityConfig:
- verify `ActiveDirectoryLdapAuthenticationProvider` does populate authorities properly (most likely, it doesn't)
- if it doesn't, copy-paste `ActiveDirectoryLdapAuthenticationProvider` 
and (re)implement a populator (unless [#4490](https://github.com/spring-projects/spring-security/issues/4490) is resolved)

via https://github.com/provectus/kafka-ui/issues/3741

### LINKED ISSUE #730: BE: RBAC: LDAP: Implement `user` subject type for LDAP & AD
Can we implement support for `user` subject type like here https://github.com/kafbat/kafka-ui/pull/719/files#diff-8a3d6c3f4c752c32a92b1b123bf39c98c4d6ee1377ad1ee1636999bdc9410a87R39 ?

Or will we do it later as a separate PR?

_Originally posted by @wernerdv in https://github.com/kafbat/kafka-ui/pull/717#discussion_r1898099369_

-----
TODO:  Implement a `user` subject type for LDAP/AD RBAC:
```
        - provider: ldap
          type: usee
          value: "haarolean"
```

### PR DESCRIPTION
**Pull request by Haarolean**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing application instances)

**What changes did you make?** (Give an overview)

**Is there anything you'd like reviewers to focus on?**
- RM TODO for #54
- skip DefaultLdapAuthoritiesPopulator bean init in case of AD (AD auth provider has the default populator we need)
- Support Provider.LDAP_AD in `RbacLdapAuthoritiesExtractor`. Resolves #54
- minor refactoring
- `user` subject type for LDAP & AD RBAC. Fixes #730

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)

- [ ] No need to
- [ ] Manually (please, describe, if necessary)
- [ ] Unit checks
- [ ] Integration checks
- [ ] Covered by existing automation

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [ ] My changes generate no new warnings (e.g. Sonar is happy)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged

Check out [Contributing](https://github.com/kafbat/kafka-ui/blob/main/.github/CONTRIBUTING.md) and [Code of Conduct](https://github.com/kafbat/kafka-ui/blob/main/.github/CODE-OF-CONDUCT.md)

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by Haarolean** [comment]
@wernerdv wanna take a look?

**Comment by wernerdv** [comment]
@Haarolean Yes, I'll check locally with these changes and report back.

**Comment by wernerdv** [comment]
@Haarolean A few comments:

1 - https://github.com/kafbat/kafka-ui/pull/717/files#diff-af0ab81f6ae459e9fd60e1979d12b70ab8556054fe012848b0e930ade92208a5R52 need `@Autowired(required = false) LdapAuthoritiesPopulator authoritiesExtractor`.
Otherwise there will be an error on startup.

2 - we can make `LdapSecurityConfig extends AbstractAuthSecurityConfig` and remove static import for `AUTH_WHITELIST`.

But most importantly, RBAC with AD still only works correctly if rbac.roles.name is group from AD.
https://github.com/kafbat/kafka-ui/blob/main/api/src/main/java/io/kafbat/ui/controller/AccessController.java#L37

If not, then /api/authorization returns empty permissions for user.

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacLdapAuthoritiesExtractor.java [hunk 1]
```diff
@@ -36,7 +38,7 @@ protected Set<GrantedAuthority> getAdditionalRoles(DirContextOperations user, St
         .stream()
         .filter(r -> r.getSubjects()
             .stream()
-            .filter(subject -> subject.getProvider().equals(Provider.LDAP))
+            .filter(subject -> SUPPORTED_PROVIDERS.contains(subject.getProvider()))
```
Comment (wernerdv) [inline api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacLdapAuthoritiesExtractor.java:41]: Do we really need this if RbacLdapAuthoritiesExtractor is not used when working with Active Directory?

File: api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java [hunk 2]
```diff
@@ -99,6 +107,10 @@ public DefaultLdapAuthoritiesPopulator ldapAuthoritiesExtractor(ApplicationConte
                                                                   AccessControlService acs) {
     var rbacEnabled = acs != null && acs.isRbacEnabled();
 
+    if (props.isActiveDirectory()) {
```
Comment (wernerdv) [inline api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java:110]: Maybe it's worth doing this check at the very beginning of the method?

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacLdapAuthoritiesExtractor.java (hunk 1 above)
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacLdapAuthoritiesExtractor.java:41]: Aaah, this makes a lot more sense now!

The reason you're encountering the issue with roles being assigned only if they match the names of AD groups is that we use the `DefaultActiveDirectoryAuthoritiesPopulator` which maps AD groups one-to-one without the matching logic we have in `RbacLdapAuthoritiesExtractor`. After all, we need some custom code, adjusting `RbacLdapAuthoritiesExtractor` to support extracting original groups from LDAP and AD.

File: api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java (hunk 2 above)
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java:110]: Sure

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacLdapAuthoritiesExtractor.java (hunk 1 above)
Comment (wernerdv) [reply api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacLdapAuthoritiesExtractor.java:41]: > After all, we need some custom code, adjusting RbacLdapAuthoritiesExtractor to support extracting original groups from LDAP and AD.

Do you need help with this?

**Comment by Haarolean** [comment]
@wernerdv hey, I've implemented the changes I've talked about here: https://github.com/kafbat/kafka-ui/pull/717#discussion_r1888542876
Additionally, did some config refactoring and fixed the stuff you suggested, PTAL again if you're interested

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacActiveDirectoryAuthoritiesExtractor.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        .map(GrantedAuthority::getAuthority)
+        .peek(group -> log.trace("Found AD group [{}] for user [{}]", group, username))
+        .collect(Collectors.toSet());
+
+    return acs.getRoles()
+        .stream()
+        .filter(r -> r.getSubjects()
+            .stream()
+            .filter(subject -> subject.getProvider().equals(Provider.LDAP_AD))
+            .filter(subject -> subject.getType().equals("group"))
```
Comment (wernerdv) [inline api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacActiveDirectoryAuthoritiesExtractor.java:39]: Can we implement support for `user` subject type like here https://github.com/kafbat/kafka-ui/pull/719/files#diff-8a3d6c3f4c752c32a92b1b123bf39c98c4d6ee1377ad1ee1636999bdc9410a87R39 ?

Or will we do it later as a separate PR?

**Comment by wernerdv** [comment]
My yaml config:
```
kafka:
  update-metrics-rate-millis: 30000 
  clusters:
    - name: LOCAL
      bootstrapServers: localhost:9092
      metrics:
        port: 7010
        type: JMX
        username: admin
        password: qwe123

auth:
  type: LDAP
spring:
  jmx:
    enabled: true
  ldap:
    urls: ldap://host:389
oauth2:
  ldap:
    activeDirectory: true
    activeDirectory.domain: <domain>

rbac:
  roles:
    - name: "my-role"
      clusters:
        - LOCAL
      subjects:
        - provider: ldap_ad
          type: group
          value: my-ad-group
      permissions:
        - resource: applicationconfig
          actions: all
        - resource: clusterconfig
          actions: all
```

Error at UI startup:
```
ERROR [main] o.s.b.SpringApplication: Application run failed
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.springframework.security.config.annotation.web.reactive.ServerHttpSecurityConfiguration': Unsatisfied dependency expressed through method 'setAuthenticationManager' parameter 0: Error creating bean with name 'authenticationManager' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Unsatisfied dependency expressed through method 'authenticationManager' parameter 0: Error creating bean with name 'authenticationProvider' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Unsatisfied dependency expressed through method 'authenticationProvider' parameter 1: Error creating bean with name 'ldapBindAuthentication' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Either an LdapUserSearch or DN pattern (or both) must be supplied.
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.resolveMethodArguments(AutowiredAnnotationBeanPostProcessor.java:896)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.inject(AutowiredAnnotationBeanPostProcessor.java:849)
        at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:145)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:509)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1439)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:599)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:975)
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:971)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:625)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:335)
        at io.kafbat.ui.KafkaUiApplication.startApplication(KafkaUiApplication.java:24)
        at io.kafbat.ui.KafkaUiApplication.main(KafkaUiApplication.java:17)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:102)
        at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:64)
        at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:40)
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'authenticationManager' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Unsatisfied dependency expressed through method 'authenticationManager' parameter 0: Error creating bean with name 'authenticationProvider' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Unsatisfied dependency expressed through method 'authenticationProvider' parameter 1: Error creating bean with name 'ldapBindAuthentication' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Either an LdapUserSearch or DN pattern (or both) must be supplied.
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:795)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:542)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredMethodElement.resolveMethodArguments(AutowiredAnnotationBeanPostProcessor.java:888)
        ... 24 common frames omitted
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'authenticationProvider' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Unsatisfied dependency expressed through method 'authenticationProvider' parameter 1: Error creating bean with name 'ldapBindAuthentication' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Either an LdapUserSearch or DN pattern (or both) must be supplied.
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:795)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:542)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:904)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:782)
        ... 37 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'ldapBindAuthentication' defined in class path resource [io/kafbat/ui/config/auth/LdapSecurityConfig.class]: Either an LdapUserSearch or DN pattern (or both) must be supplied.
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1806)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:904)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:782)
        ... 51 common frames omitted
Caused by: java.lang.IllegalArgumentException: Either an LdapUserSearch or DN pattern (or both) must be supplied.
        at org.springframework.util.Assert.isTrue(Assert.java:111)
        at org.springframework.security.ldap.authentication.AbstractLdapAuthenticator.afterPropertiesSet(AbstractLdapAuthenticator.java:71)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1853)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1802)
        ... 62 common frames omitted
```

File: api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java [hunk 4]
```diff
@@ -43,44 +42,55 @@
...
+    if (rbacEnabled) {
+      authProvider.setUserDetailsContextMapper(new RbacUserDetailsMapper());
+    }
+
+    return authProvider;
+  }
+
+  @Bean
+  public BindAuthenticator ldapBindAuthentication(LdapContextSource ldapContextSource) {
     BindAuthenticator ba = new BindAuthenticator(ldapContextSource);
```
Comment (wernerdv) [inline api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java:80]:
```
if (props.isActiveDirectory()) {
  return null;
}
```

File: api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java [hunk 5]
```diff
@@ -43,44 +42,55 @@
...
-  public ReactiveAuthenticationManager authenticationManager(LdapContextSource ldapContextSource,
-                                                             LdapAuthoritiesPopulator authoritiesExtractor,
-                                                             AccessControlService acs) {
+  public ReactiveAuthenticationManager authenticationManager(AbstractLdapAuthenticationProvider authProvider) {
+    return new ReactiveAuthenticationManagerAdapter(new ProviderManager(List.of(authProvider)));
+  }
+
+  @Bean
+  public AbstractLdapAuthenticationProvider authenticationProvider(LdapAuthoritiesPopulator authoritiesExtractor,
+                                                                   BindAuthenticator bindAuthenticator,
```
Comment (wernerdv) [inline api/src/main/java/io/kafbat/ui/config/auth/LdapSecurityConfig.java:56]: `@Autowired(required=false) BindAuthenticator bindAuthenticator`

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacActiveDirectoryAuthoritiesExtractor.java (hunk 3 above)
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacActiveDirectoryAuthoritiesExtractor.java:39]: Oh, I didn't even think about the fact that we don't have a way to do this for LDAP/AP.

Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/service/rbac/extractor/RbacActiveDirectoryAuthoritiesExtractor.java:39]: Will do this within this PR, but raised #730 for the changelog transparency

**Comment by Haarolean** [comment]
> Otherwise there will be an error on startup.

@wernerdv thanks, fixed!

**Review by wernerdv — COMMENTED** [review]
I checked and everything works as expected

**Comment by Haarolean** [comment]
@wernerdv thanks for the review and trying out the changes!

**Pull request merged**